src/renderers/shaders/ShaderChunk/transparent_fragment.glsl.js

```diff
@@ -0,0 +1,7 @@
+export default /* glsl */`
+#ifndef TRANSPARENT
+
+	diffuseColor.a = 1.0;
```

Comment:
`.transparent === false` only disables blending if `.blending` is `NormalBlending`.

See #14171.

So with custom blending, blending is still enabled even if .`transparent` is `false`.

You are changing the user's data. I'm not sure about this change -- so, raising a red flag...

---

Also, if you choose to retain this, then I'd call this file `opaque_fragment.glsl.js` since that is what it applies to.

/ping @mrdoob @donmccurdy

Reply:
Confirmed. This breaks `AdditiveBlending` when `texture.a` < 1 and `material.transparent` is `false`.

Reply:
Good points, thanks — I'd assumed `.transparent = false` implied that all blending was disabled.

One short-term fix would be to treat the material as if `.transparent === true` automatically, if any non-default blend mode is set. That sounds like it would cover the intended uses of the API, or at least I don't know why the user would *intentionally* use `.transparent = false` and a non-default blend mode? However, we are short on time to test a change like that before r114, and I'm not sure I understand the conceptual differences between:

- `.transparent = true` vs. `.blending = NormalBlending`
- `.transparent = false` vs. `.blending = NoBlending`

I think some changes to the transparency API might be worth considering later, but we certainly can't get that done in time for r114...

Reply:
What about https://github.com/mrdoob/three.js/pull/14171#issuecomment-397846606? Was this feedback ever honored? Would it be also a solution to revert #14171?

Reply:
I'm afraid I don't understand what is meant in https://github.com/mrdoob/three.js/pull/14171 by "clearly not correct", or what transparency-related problems @jbaicoianu was referring to. These other blending modes seem (to me) to fit under the umbrella of semitransparency, and I would think you do want the sorting behavior that `.transparent=true` provides, even for other blend modes.

@jbaicoianu do you remember what problems you were referring to?

Reply:
It's been a while and I think for me that issue was more of an "ah hah, so THAT'S why my blend modes weren't being applied consistently!" kind of thing. I think after reading #14171 I just modified my engine to always force ```transparent=true``` if a different blend mode was set by the user

Reply:
>I'm afraid I don't understand what is meant in #14171 by "clearly not correct"...

For example, `AdditiveBlending` can be used with opaque materials. It does not make sense in that case for a user to have to set `transparent = true` to properly enable `AdditiveBlending`.

>Would it be also a solution to revert #14171?

In my current thinking about this issue, no.

Reply:
>These other blending modes seem (to me) to fit under the umbrella of semitransparency,

I don't think I would say that. Transparency is associated with one, particular, blending mode: `NormalBlending` -- the mode that best models the behavior of light passing through translucent materials.

>and I would think you do want the sorting behavior that .transparent=true provides, even for other blend modes.

Sorting may be required in certain cases, but users can set `depthWrite = false` and/or `renderOrder` for that.

Reply:
Rendering front-to-back alongside opaque objects defeats the purpose of adding to the color behind it, no? In [Unreal](https://docs.unrealengine.com/en-US/Resources/ContentExamples/MaterialProperties/1_1/index.html), for example, "opaque" is considered a blend mode, and mutually exclusive with additive or translucent blend modes.

Regardless, if r114 is still planned to go out this week, I agree with @WestLangley that this needs a quick fix or a temporary revert of this PR (#18631) and #18579.

Reverting #14171 might be an option for r115, but is more than I (at least personally) have time to consider the full consequences of this week.

Reply:
>Rendering front-to-back alongside opaque objects defeats the purpose of adding to the color behind it, no?

Right. For additive blending, the order does not matter, but `depthWrite = false` would be required.

Reply:
> I don't think I would say that. Transparency is associated with one, particular, blending mode: `NormalBlending` -- the mode that best models the behavior of light passing through translucent materials.

My current understanding of `material.transparent` is unrelated to blending. It's related to ignoring (or not) the alpha channels in `map` and/or `alphaMap`.

I'm not sure if this thinking makes things easier or not.

Reply:
>My current understanding of material.transparent is unrelated to blending. It's related to ignoring (or not) the alpha channels in map and/or alphaMap.

`material.transparent = true` is used as a signal to the renderer to render the object in a later pass, sorted from far to near. There is no other way for the renderer to know to do that without the user setting a flag. `transparent` is a reasonable name for that flag, but it could have been named something more descriptive.

`material.transparent = false` is used to disable blending when `NormalBlending` is set, but doing so is just for efficiency's sake.

//

If you want to ignore -- or not ignore -- the alpha channel,  `alphaMode` could be used for that.

Reply:
Ah, so you're suggesting we add a `alphaMode` property to `material` that follows GLTF options?

I guess the options would then be `opaque`, `blend` and `mask`. And in our case `blend` would be default?

Reply:
I think `.alphaMode` — if implemented in addition to `.transparent` and `.blending` — would become confusing. Also, glTF's constraint that blending and masking be mutually exclusive isn't necessarily something  we want ... although Blender does the same, so maybe it's fine?

Something like `.alphaMode` could be a replacement for `.transparent`, though. Consider:

```
material.alphaMode = Opaque | Blend | Dither | ... ;
material.alphaBlend = NormalBlend | AdditiveBlend | CustomBlend | ... ;
material.alphaTest = <number>;
```

In that case `.alphaBlend` has an effect only when `.alphaMode = Blend`. That mode would also force the render order currently associated with `.transparent=true`. Then `.alphaTest` is independent, and can be used in any mode. This covers all of glTF's API, and would support future addition of alpha dither or alpha hash modes. It doesn't cover 100% of three.js' current API (can't do `.transparent=false` + additive blending) but it covers all of the three.js API that I know of a use for. 

There's also [physically-based transparency](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/ADOBE_materials_thin_transparency) to think of, but my current assumption is that it would remain a separate — more advanced — API, available only on MeshPhysicalMaterial.

Reply:
Yes.

I would deprecate `alphaTest` and replace it with `.alphaMode` and `.alphaCutoff`.

I assume `.alphaMode` would be implemented as a `#DEFINE`; `.alphaCutoff` as a uniform.

Reply:
>Ah, so you're suggesting we add a alphaMode property to material that follows GLTF options?

Yes, that was what I was thinking. Deprecate `alphaTest`, and replace it with `.alphaMode` and `.alphaCutoff`.

//

However, something similar to @donmccurdy's suggestion may also work:

    material.alphaMode = Opaque | Blend | Mask | ... ;
    material.alphaBlend = NormalBlend | AdditiveBlend | CustomBlend | ... ;
    material.alphaCutoff = <number>;

and deprecate `transparent`, `blending`, and `alphaTest`.

Reply:
To unblock r114, how about this: we detect when `.transparent === false` and `.blending > NormalBlending`, logging a warning:

> THREE.WebGLRenderer: Setting .transparent=true is required when using non-opaque .blending modes.

A combination of `.transparent=false` and `.blending=AdditiveBlending` will still behave differently than it has in previous versions, but:

1. Users can easily see if they've hit this problem.
2. If they didn't have a particular reason for using `.transparent=false`, it's easy to fix.
3. If they *did* have a particular reason for using `.transparent=false`, we'll get that feedback before we invest time in anything like `.alphaMode` that would couple transparent sorting to the other blend modes.

Reply:
How about if we revert this PR (and related PRs) and try again at the start of the next cycle.

Reply:
I'd be fine with that as well.

Reply:
In the mean time, I found Unity's use of render queues very interesting: https://docs.unity3d.com/Manual/SL-SubShaderTags.html.

Reply:
Okay, reverting.

Reply:
<img width="935" alt="Screen Shot 2020-02-29 at 1 27 47 PM" src="https://user-images.githubusercontent.com/97088/75615352-5a804880-5af7-11ea-9c12-fd4dc792cffb.png">

Uh oh...

Reply:
Lets see if I can do it locally...

Reply:
@mrdoob Also see https://github.com/mrdoob/three.js/pull/18631#discussion_r383594673.

Reply:
> @mrdoob Also see #18631 (comment).

I'm too confused now. What about it?

Reply:
Reverted: https://github.com/mrdoob/three.js/commit/4361a31b1ad3c4b6021d6de463569e126e196df4 https://github.com/mrdoob/three.js/commit/8582c7437925ab3f23ed2d5a43d21e254ede5fa5

Reply:
He referred to #18579, too.

Reply:
Oh lord. Yes, just noticed...
https://github.com/mrdoob/three.js/pull/18579/files#diff-a47375d4b502ebcbbf70db4674af6724R5

Reply:
Okay, reverted  #18579 too. Thanks! 🙏